dotnet/Query/Query/Query.cs

```diff
@@ -0,0 +1,558 @@
+
+    internal void UpdateScenario(Scenario scenario)
+    {
+        UpdateSteps(scenario.Steps);
+    }
+
+    internal void UpdateSteps(IList<Step> steps)
+    {
+        foreach (var step in steps)
+        {
+            _stepById[step.Id] = step;
+        }
+    }
+
+    internal void UpdateMeta(Meta meta)
+    {
+        _meta = meta;
+    }
+
+    public void Update(Envelope envelope)
+    {
+        if (envelope.Meta != null) UpdateMeta(envelope.Meta);
+        if (envelope.TestRunStarted != null) UpdateTestRunStarted(envelope.TestRunStarted);
+        if (envelope.TestRunFinished != null) UpdateTestRunFinished(envelope.TestRunFinished);
+        if (envelope.TestCaseStarted != null) UpdateTestCaseStarted(envelope.TestCaseStarted);
+        if (envelope.TestCaseFinished != null) UpdateTestCaseFinished(envelope.TestCaseFinished);
+        if (envelope.TestStepStarted != null) UpdateTestStepStarted(envelope.TestStepStarted);
+        if (envelope.TestStepFinished != null) UpdateTestStepFinished(envelope.TestStepFinished);
+        if (envelope.GherkinDocument != null) UpdateGherkinDocument(envelope.GherkinDocument);
+        if (envelope.Pickle != null) UpdatePickle(envelope.Pickle);
+        if (envelope.TestCase != null) UpdateTestCase(envelope.TestCase);
+        if (envelope.Hook != null) UpdateHook(envelope.Hook);
+        if (envelope.Attachment != null) UpdateAttachment(envelope.Attachment);
+    }
+
+    // FindLineageBy methods
+    public Lineage? FindLineageBy(GherkinDocument element)
+    {
+        _lineageById.TryGetValue(element, out var lineage);
+        return lineage;
+    }
+
+    public Lineage? FindLineageBy(Feature element)
+    {
+        _lineageById.TryGetValue(element, out var lineage);
+        return lineage;
+    }
+
+    public Lineage? FindLineageBy(Rule element)
+    {
+        _lineageById.TryGetValue(element, out var lineage);
+        return lineage;
+    }
+
+    public Lineage? FindLineageBy(Scenario element)
+    {
+        _lineageById.TryGetValue(element, out var lineage);
+        return lineage;
+    }
+
+    public Lineage? FindLineageBy(Examples element)
+    {
+        _lineageById.TryGetValue(element, out var lineage);
+        return lineage;
+    }
+
+    public Lineage? FindLineageBy(TableRow element)
+    {
+        _lineageById.TryGetValue(element, out var lineage);
+        return lineage;
+    }
+
+    public Lineage? FindLineageBy(Pickle pickle)
+    {
+        var astNodeIds = pickle.AstNodeIds;
+        var lastAstNodeId = astNodeIds.LastOrDefault();
+        _lineageById.TryGetValue(lastAstNodeId, out var lineage);
+        return lineage;
+    }
+
+    public Lineage? FindLineageBy(TestCaseStarted testCaseStarted)
+    {
+        var pickle = FindPickleBy(testCaseStarted);
+        if (pickle == null)
+        {
+            return null;
+        }
+        return FindLineageBy(pickle);
+    }
+
+    public string FindNameOf(GherkinDocument element, NamingStrategy namingStrategy)
```

Comment:
@davidjgoss since we made the lineage API public, I think we could get rid of all the `FindNameOf` methods.

Reply:
Made https://github.com/cucumber/query/pull/84 to show.